slick/src/main/scala/slick/jdbc/DriverDataSource.scala

```diff
@@ -46,42 +46,45 @@ class DriverDataSource(
   // State that gets initialized by `init`
   @volatile private[this] var registered: Boolean = false
   @volatile private[this] var initialized = false
-  private[this] var driver: Driver = _
-  private[this] var connectionProps: Properties = _
+  @volatile private[this] var driver: Driver = _
+  @volatile private[this] var connectionProps: Properties = _
```

Comment:
In the two `def getConnection` methods they are read without any synchronization.

```scala
  def getConnection: Connection = {
    init
    driver.connect(url, connectionProps)
  }

  def getConnection(username: String, password: String): Connection = {
    init
    driver.connect(url, propsWithUserAndPassword(connectionProps, username, password))
  }
```
https://github.com/mustajarvi/slick/blob/fix-%231400-npe-in-init/slick/src/main/scala/slick/jdbc/DriverDataSource.scala#L104

Unless the reads are also synchronized in some way there's nothing preventing the CPU from just reusing stale values if those happen to exist in L2/L1 caches. Volatile will prevent this.